app/models/developer.rb

```diff
@@ -25,6 +25,7 @@ class Developer < ApplicationRecord
     max_file_size: 10.megabytes
 
   scope :filter_by_utc_offset, ->(utc_offset) { where(utc_offset:) }
+  scope :filter_by_role_types, ->(role_types) { joins(:role_type).where("part_time_contract=? or full_time_contract=? or full_time_employment=?", role_types.include?(:part_time_contract), role_types.include?(:full_time_contract), role_types.include?(:full_time_employment)) }
```

Comment:
I don't think this is querying for what we want.

If I don't select `full_time_contract` then this query will filter out developers who _aren't_ looking for a full-time contract. Instead, it should be including developers who both are and aren't looking. Does that make sense?

Said another way, each of these three statements should be added conditionally if the value is selected.

Reply:
I have a confusion regarding your comment. Here's how the current implementation will work:

- If I select `full_time_contract` then this query will output developers who are looking for a full-time contact only.
- If I select `full_time_contract` and `part_time_contract` then this query will output developers who are either looking for a full-time contract or part-time contract, and same with any other combination.
- and if I select all three options then this query will output developers who are looking for a full-time contract or a part-time contract or a full-time employment.

Can you please tell which developers to output in each of the above scenario? So we are on the same page. Thanks.

Reply:
* If I select full-time then I should only see developers that have full-time checked.
* If I select full-time and part-time I should only see developers that have one or the other checked.
* If I select all three then I should see developers with any of the three options checked.
* If I select none then the filter isn't applied and all developers are shown.

The last one is working, so we can ignore that. But if I output the SQL from the first one here is the query.

```ruby
[12] pry(main)> Developer.filter_by_role_types([:full_time_contract]).to_sql
=> "SELECT \"developers\".* FROM \"developers\" INNER JOIN \"role_types\" ON \"role_types\".\"developer_id\" = \"developers\".\"id\" WHERE (part_time_contract=FALSE or full_time_contract=TRUE or full_time_employment=FALSE)"
```

That means we are lumping in folks who set the other two to `false`. **This is including folks who have `false` set for all three values in all queries.***

Compare that to this query, which only sets one of the values:

```ruby
[28] pry(main)> Developer.joins(:role_type).where(role_type: {part_time_contract: true}).to_sql
=> "SELECT \"developers\".* FROM \"developers\" INNER JOIN \"role_types\" \"role_type\" ON \"role_type\".\"developer_id\" = \"developers\".\"id\" WHERE \"role_type\".\"part_time_contract\" = TRUE"
```

Building on that, if someone filtered by more than one value:

```ruby
[36] pry(main)> Developer.joins(:role_type).where(role_type: {part_time_contract: true}).or(Developer.joins(:role_type).where(role_type: {full_time_contract: true})).to_sql
=> "SELECT \"developers\".* FROM \"developers\" INNER JOIN \"role_types\" \"role_type\" ON \"role_type\".\"developer_id\" = \"developers\".\"id\" WHERE (\"role_type\".\"part_time_contract\" = TRUE OR \"role_type\".\"full_time_contract\" = TRUE)"
```

---

Does that make sense?

Reply:
I think that this should get us what we want.

```suggestion
  scope :filter_by_role_types, ->(role_types) do
    RoleType::TYPES.filter_map { |type|
      where(role_type: {type => true}) if role_types.include?(type)
    }.reduce(:or).joins(:role_type)
  end
```

```ruby
[64] pry(main)> Developer.filter_by_role_types([:part_time_contract, :full_time_contract]).to_sql
=> "SELECT \"developers\".* FROM \"developers\" INNER JOIN \"role_types\" \"role_type\" ON \"role_type\".\"developer_id\" = \"developers\".\"id\" WHERE (\"role_type\".\"part_time_contract\" = TRUE OR \"role_type\".\"full_time_contract\" = TRUE)"
```

Reply:
Yeah that makes sense. Thanks. I'll update the PR.